File: backend/src/fplai/optimizer/rating.py
"""Rate an arbitrary 15-man squad against the dream-team benchmark (stage-5 support).

:func:`rate_team` scores a submitted squad over the live prediction window with a pure,
deterministic metric::

    metric(squad) = sum over the window's GWs of (best-legal-XI xP sum + captain bonus)

The best XI for a GW is chosen greedily — formation minimums (1 GKP/3 DEF/2 MID/1 FWD)
by xP first, then filled to 11 by xP respecting the maxes (1 GKP/5 DEF/5 MID/3 FWD) —
and the captain bonus is the highest xP within that XI.  The score normalises the
metric between two anchors evaluated with the SAME metric::

    score = clamp01((metric(team) - metric(floor)) / (metric(optimal) - metric(floor))) * 100

``optimal`` is the squad from ``data/processed/dream_team.json`` and ``floor`` is the
cheapest legal 15 from the live roster.  A user squad CAN outscore the dream team on
this metric — the dream team is solved against a decayed, chip-aware, bench-weighted
MILP objective over its own horizon, not this undecayed XI-sum — so the score is
clamped to 100 rather than treated as an error.

Everything here is pure and deterministic: no I/O, no clock, every tie broken by
``player_code``.  Players with no prediction rows in the window contribute 0 xP but
remain legal picks (e.g. brand-new signings the model has not scored yet); ``q0`` is
taken from the player's first prediction row within the window, where present.
"""

from __future__ import annotations

from collections import Counter
from typing import TYPE_CHECKING, Literal, NamedTuple

import pandas as pd
from pydantic import BaseModel, Field

from fplai import rules
from fplai.optimizer.milp import (
    CLUB_LIMIT,
    FORMATION_MAX,
    FORMATION_MIN,
    INITIAL_BUDGET,
    POSITION_QUOTA,
    SQUAD_SIZE,
    XI_SIZE,
)

if TYPE_CHECKING:
    from collections.abc import Iterable, Mapping, Sequence

__all__ = [
    "PlayerRating",
    "RatingResult",
    "TeamValidationError",
    "VERDICT_BANDS",
    "WeakestPlayer",
    "XiSelection",
    "best_xi",
    "cheapest_legal_squad",
    "rate_team",
    "squad_metric",
    "validate_squad",
]

#: Score bands, checked in order: first threshold the score reaches wins.
VERDICT_BANDS: tuple[tuple[float, str], ...] = (
    (95.0, "ELITE"),
    (85.0, "STRONG"),
    (70.0, "SOLID"),
    (50.0, "ROUGH"),
)

Verdict = Literal["ELITE", "STRONG", "SOLID", "ROUGH", "FODDER"]


class TeamValidationError(ValueError):
    """The submitted squad is illegal; ``problems`` lists EVERY violated rule."""

    def __init__(self, problems: list[str]) -> None:
        self.problems = problems
        super().__init__("; ".join(problems))


# --------------------------------------------------------------------------------------
# Response models (mirror the POST /api/rate-team contract field-for-field)
# --------------------------------------------------------------------------------------


class PlayerRating(BaseModel):
    """One squad member's contribution over the rating window."""

    player_code: int
    web_name: str
    position: str
    team_short: str | None = None
    price: int = Field(description="Current price, £0.1m units")
    xp_gw1: float = Field(description="xP at the window's first GW (0 when no row)")
    xp_horizon: float = Field(description="Summed xP over the window's GWs")
    q0: float | None = Field(
        default=None, description="P(0 minutes) from the player's first window row"
    )
    in_best_xi_gw1: bool


class WeakestPlayer(BaseModel):
    """One of the (up to 3) lowest-xP squad members over the window."""

    player_code: int
    web_name: str
    xp_horizon: float


class RatingResult(BaseModel):
    """POST /api/rate-team response: the squad's score against the benchmark anchors.

    ``team_xp_gw1`` is the first window GW's metric term (best-XI xP + captain bonus);
    ``team_xp_horizon`` is the full-window metric, so the per-GW terms sum to it.
    """

    score: float = Field(ge=0.0, le=100.0)
    verdict: Verdict
    season: int
    from_gw: int
    horizon: int = Field(description="Number of GWs in the rating window")
    team_xp_gw1: float
    team_xp_horizon: float
    optimal_xp_horizon: float = Field(description="metric(dream-team squad), same metric")
    floor_xp_horizon: float = Field(description="metric(cheapest legal 15), same metric")
    best_xi_gw1: list[int]
    formation_gw1: str = Field(description="DEF-MID-FWD, e.g. '3-5-2'")
    suggested_captain: int = Field(description="Highest-xP member of the first GW's XI")
    player_ratings: list[PlayerRating]
    weakest: list[WeakestPlayer]


class XiSelection(NamedTuple):
    """A greedy best-XI pick for one GW."""

    xi: list[int]
    formation: str
    xi_xp: float
    captain: int
    captain_xp: float


# --------------------------------------------------------------------------------------
# Squad legality
# --------------------------------------------------------------------------------------


def validate_squad(player_codes: Sequence[int], roster: pd.DataFrame) -> list[str]:
    """Every squad-legality rule the submitted codes violate (empty = legal).

    Rules per FPL_KNOWLEDGE §1.3: exactly 15 distinct players, quotas 2 GKP/5 DEF/
    5 MID/3 FWD, at most 3 per club, total price within the £100m budget, and every
    code present in the live roster.  All checks run — nothing short-circuits — so
    the caller can report every violation at once.
    """
    problems: list[str] = []
    codes = [int(c) for c in player_codes]
    if len(codes) != SQUAD_SIZE:
        problems.append(f"need exactly {SQUAD_SIZE} players, got {len(codes)}")
    for code in sorted(c for c, n in Counter(codes).items() if n > 1):
        problems.append(f"duplicate player_code {code}")

    info: dict[int, tuple[str, int, int]] = {
        int(r.player_code): (str(r.position), int(r.team_code), int(r.price))
        for r in roster.itertuples(index=False)
    }
    known = [c for c in codes if c in info]
    for code in sorted(set(codes) - set(known)):
        problems.append(f"unknown player_code {code}")

    counts = Counter(info[c][0] for c in known)
    for pos in rules.POSITIONS:
        if counts.get(pos, 0) != POSITION_QUOTA[pos]:
            problems.append(
                f"position quota: need {POSITION_QUOTA[pos]} {pos}, got {counts.get(pos, 0)}"
            )
    for club, n in sorted(Counter(info[c][1] for c in known).items()):
        if n > CLUB_LIMIT:
            problems.append(
                f"club limit: max {CLUB_LIMIT} players per club, got {n} from team_code {club}"
            )
    total_price = sum(info[c][2] for c in known)
    if total_price > INITIAL_BUDGET:
        problems.append(f"total price {total_price} exceeds budget {INITIAL_BUDGET}")
    return problems


# --------------------------------------------------------------------------------------
# The metric: greedy best XI + captain bonus, summed over the window
# --------------------------------------------------------------------------------------


def best_xi(entries: Iterable[tuple[int, str, float]]) -> XiSelection:
    """Greedy best legal XI from ``(player_code, position, xp)`` entries.

    Formation minimums (1 GKP/3 DEF/2 MID/1 FWD) by xP first, then filled to 11 by
    xP respecting the maxes (1 GKP/5 DEF/5 MID/3 FWD).  Every tie breaks by
    ``player_code`` for determinism; the captain is the XI's highest-xP player.

    Raises:
        ValueError: when the entries cannot field a formation-legal XI.
    """
    by_pos: dict[str, list[tuple[int, float]]] = {pos: [] for pos in rules.POSITIONS}
    for code, pos, xp in entries:
        if pos not in by_pos:
            raise ValueError(f"unknown position {pos!r} for player {code}")
        by_pos[pos].append((int(code), float(xp)))
    for group in by_pos.values():
        group.sort(key=lambda e: (-e[1], e[0]))

    chosen: list[tuple[int, str, float]] = []
    counts: dict[str, int] = {}
    for pos in rules.POSITIONS:
        need = FORMATION_MIN[pos]
        if len(by_pos[pos]) < need:
            raise ValueError(f"cannot field a legal XI: need {need} {pos}, have {len(by_pos[pos])}")
        chosen.extend((code, pos, xp) for code, xp in by_pos[pos][:need])
        counts[pos] = need
    rest = sorted(
        (
            (code, pos, xp)
            for pos in rules.POSITIONS
            for code, xp in by_pos[pos][FORMATION_MIN[pos] :]
        ),
        key=lambda e: (-e[2], e[0]),
    )
    for code, pos, xp in rest:
        if len(chosen) == XI_SIZE:
            break
        if counts[pos] >= FORMATION_MAX[pos]:
            continue
        chosen.append((code, pos, xp))
        counts[pos] += 1
    if len(chosen) < XI_SIZE:
        raise ValueError(f"cannot field a legal XI: only {len(chosen)} eligible players")

    captain, captain_xp = max(((c, xp) for c, _, xp in chosen), key=lambda e: (e[1], -e[0]))
    order = {pos: i for i, pos in enumerate(rules.POSITIONS)}
    xi = [c for c, _, _ in sorted(chosen, key=lambda e: (order[e[1]], -e[2], e[0]))]
    return XiSelection(
        xi=xi,
        formation=f"{counts['DEF']}-{counts['MID']}-{counts['FWD']}",
        xi_xp=sum(xp for _, _, xp in chosen),
        captain=captain,
        captain_xp=captain_xp,
    )


def squad_metric(
    codes: Sequence[int],
    pos_of: Mapping[int, str],
    xp_of: Mapping[tuple[int, int], float],
    window: Iterable[int],
) -> float:
    """``sum over window GWs of (best-XI xP + captain bonus)`` for one 15-man squad.

    ``xp_of`` maps ``(gw, player_code) -> xp``; players without a row contribute 0.
    """
    total = 0.0
    for gw in window:
        sel = best_xi((c, pos_of[c], xp_of.get((gw, c), 0.0)) for c in codes)
        total += sel.xi_xp + sel.captain_xp
    return total


def cheapest_legal_squad(roster: pd.DataFrame) -> list[int]:
    """The floor squad: cheapest quota-satisfying, club-legal 15 from the roster.

    Greedy per position (cheapest first, ties by ``player_code``), skipping players
    whose club is already at the limit — the same warm-start greedy the MILP uses.

    Raises:
        ValueError: when the roster cannot field a legal 15.
    """
    chosen: list[int] = []
    club_n: Counter[int] = Counter()
    for pos, quota in POSITION_QUOTA.items():
        group = roster[roster["position"] == pos].sort_values(
            ["price", "player_code"], kind="stable"
        )
        taken = 0
        for row in group.itertuples(index=False):
            if club_n[int(row.team_code)] >= CLUB_LIMIT:
                continue
            chosen.append(int(row.player_code))
            club_n[int(row.team_code)] += 1
            taken += 1
            if taken == quota:
                break
        if taken < quota:
            raise ValueError(f"roster cannot field a legal 15: only {taken} {pos} available")
    return chosen


# --------------------------------------------------------------------------------------
# rate_team
# --------------------------------------------------------------------------------------


def _clamp01(x: float) -> float:
    return max(0.0, min(1.0, x))


def _verdict(score: float) -> Verdict:
    for threshold, verdict in VERDICT_BANDS:
        if score >= threshold:
            return verdict  # type: ignore[return-value]
    return "FODDER"


def rate_team(
    player_codes: Sequence[int],
    predictions_gw: pd.DataFrame,
    roster: pd.DataFrame,
    dream_squad: list[int],
    *,
    season: int,
    from_gw: int,
    horizon: int,
) -> RatingResult:
    """Score a 15-man squad 0-100 between the floor and dream-team anchors.

    Pure and deterministic — see the module docstring for the metric and the
    clamp-to-100 rationale.  ``roster`` is the legality pool (an optional
    ``team_short`` column feeds the per-player ratings); ``predictions_gw`` needs
    ``season, gw, player_code, xp`` (+ ``q0`` where available).

    Raises:
        TeamValidationError: with EVERY violated legality rule (API turns this
            into the contract's 422 ``{"detail": [...]}``).
        ValueError: when the dream squad or roster cannot be evaluated at all.
    """
    problems = validate_squad(player_codes, roster)
    if problems:
        raise TeamValidationError(problems)
    codes = [int(c) for c in player_codes]
    window = range(from_gw, from_gw + horizon)

    preds = predictions_gw[
        (predictions_gw["season"] == season) & predictions_gw["gw"].isin(window)
    ]
    xp_of: dict[tuple[int, int], float] = {
        (int(gw), int(code)): float(xp)
        for gw, code, xp in zip(preds["gw"], preds["player_code"], preds["xp"], strict=True)
    }
    q0_of: dict[int, float] = {}
    if "q0" in preds.columns:
        first_rows = preds.sort_values(["player_code", "gw"], kind="stable")
        for row in first_rows.itertuples(index=False):
            code = int(row.player_code)
            if code not in q0_of and pd.notna(row.q0):
                q0_of[code] = float(row.q0)

    pos_of: dict[int, str] = {
        int(r.player_code): str(r.position) for r in roster.itertuples(index=False)
    }
    missing = [c for c in dream_squad if c not in pos_of]
    if missing and "position" in preds.columns:
        for row in preds[preds["player_code"].isin(missing)].itertuples(index=False):
            pos_of.setdefault(int(row.player_code), str(row.position))
        missing = [c for c in dream_squad if c not in pos_of]
    if missing:
        raise ValueError(f"dream squad players unknown to roster/predictions: {missing}")

    team_metric = squad_metric(codes, pos_of, xp_of, window)
    optimal_metric = squad_metric([int(c) for c in dream_squad], pos_of, xp_of, window)
    floor_metric = squad_metric(cheapest_legal_squad(roster), pos_of, xp_of, window)

    denom = optimal_metric - floor_metric
    if denom > 0:
        score = _clamp01((team_metric - floor_metric) / denom) * 100.0
    else:  # degenerate anchors (e.g. empty window): fall back to a pass/fail split
        score = 100.0 if team_metric >= optimal_metric else 0.0

    sel_gw1 = best_xi((c, pos_of[c], xp_of.get((from_gw, c), 0.0)) for c in codes)
    xi_gw1 = set(sel_gw1.xi)

    xp_horizon_of = {c: sum(xp_of.get((gw, c), 0.0) for gw in window) for c in codes}
    detail: dict[int, tuple[str, str | None, int]] = {}
    for row in roster.itertuples(index=False):
        code = int(row.player_code)
        if code in xp_horizon_of:
            short = getattr(row, "team_short", None)
            detail[code] = (
                str(row.web_name),
                str(short) if pd.notna(short) else None,
                int(row.price),
            )
    ratings = [
        PlayerRating(
            player_code=c,
            web_name=detail[c][0],
            position=pos_of[c],
            team_short=detail[c][1],
            price=detail[c][2],
            xp_gw1=xp_of.get((from_gw, c), 0.0),
            xp_horizon=xp_horizon_of[c],
            q0=q0_of.get(c),
            in_best_xi_gw1=c in xi_gw1,
        )
        for c in codes
    ]
    weakest = [
        WeakestPlayer(player_code=c, web_name=detail[c][0], xp_horizon=xp_horizon_of[c])
        for c in sorted(codes, key=lambda c: (xp_horizon_of[c], c))[:3]
    ]

    return RatingResult(
        score=score,
        verdict=_verdict(score),
        season=season,
        from_gw=from_gw,
        horizon=horizon,
        team_xp_gw1=sel_gw1.xi_xp + sel_gw1.captain_xp,
        team_xp_horizon=team_metric,
        optimal_xp_horizon=optimal_metric,
        floor_xp_horizon=floor_metric,
        best_xi_gw1=sel_gw1.xi,
        formation_gw1=sel_gw1.formation,
        suggested_captain=sel_gw1.captain,
        player_ratings=ratings,
        weakest=weakest,
    )

File: frontend/src/lib/rating.ts
/**
 * Client-side AI-Rating engine — an EXACT pure-TS port of the Python metric in
 * backend/src/fplai/optimizer/rating.py (rate_team + best_xi + squad_metric +
 * validate_squad). Static (public) builds score squads in the browser from the
 * published bundle (players.json + xp.json + rating.json) with this engine;
 * local builds keep POST /api/rate-team.
 *
 *     metric(squad) = sum over the window's GWs of (best-legal-XI xP sum + captain bonus)
 *     score = clamp01((metric(team) - floor) / (optimal - floor)) * 100
 *
 * The best XI for a GW is chosen greedily — formation minimums (1 GKP/3 DEF/
 * 2 MID/1 FWD) by xP first, then filled to 11 by xP respecting the maxes
 * (1 GKP/5 DEF/5 MID/3 FWD) — and the captain bonus is the highest xP within
 * that XI. `floor`/`optimal` anchors arrive precomputed in rating.json (same
 * metric, evaluated by the Python engine over the same window). Every tie
 * breaks by player_code, mirroring the Python engine bit for bit.
 *
 * This module is dependency-free and framework-free on purpose: `node --test`
 * runs the parity suite (tests/rating.parity.test.ts) against fixtures exported
 * from the real Python engine.
 */

// --------------------------------------------------------------------------------------
// Constants — mirrors of fplai.rules.POSITIONS + fplai.optimizer.milp squad constants
// --------------------------------------------------------------------------------------

export const RATING_POSITIONS = ["GKP", "DEF", "MID", "FWD"] as const;
export type RatingPosition = (typeof RATING_POSITIONS)[number];

export const SQUAD_SIZE = 15;
export const XI_SIZE = 11;
export const POSITION_QUOTA: Record<RatingPosition, number> = { GKP: 2, DEF: 5, MID: 5, FWD: 3 };
export const FORMATION_MIN: Record<RatingPosition, number> = { GKP: 1, DEF: 3, MID: 2, FWD: 1 };
export const FORMATION_MAX: Record<RatingPosition, number> = { GKP: 1, DEF: 5, MID: 5, FWD: 3 };
export const CLUB_LIMIT = 3;
export const INITIAL_BUDGET = 1000; // £100.0m in £0.1m units

/** Score bands, checked in order: first threshold the score reaches wins. */
export const VERDICT_BANDS: readonly (readonly [number, RatingVerdict])[] = [
  [95.0, "ELITE"],
  [85.0, "STRONG"],
  [70.0, "SOLID"],
  [50.0, "ROUGH"],
];

export type RatingVerdict = "ELITE" | "STRONG" | "SOLID" | "ROUGH" | "FODDER";

/** Thrown by rateTeam on an illegal squad; `problems` lists EVERY violated rule. */
export class RatingValidationError extends Error {
  readonly problems: string[];

  constructor(problems: string[]) {
    super(problems.join("; "));
    this.name = "RatingValidationError";
    this.problems = problems;
  }
}

// --------------------------------------------------------------------------------------
// Inputs — the bundle shapes the engine consumes (players.json / xp.json / rating.json)
// --------------------------------------------------------------------------------------

/** One players.json row (the legality pool). */
export interface RatingPoolPlayer {
  player_code: number;
  web_name: string;
  position: string;
  team_short?: string | null;
  team_code: number;
  price: number; // £0.1m units
  q0_gw1?: number | null;
}

/** xp.json: compact per-GW horizon table; players["<code>"][i] pairs with gws[i]. */
export interface RatingXpTable {
  gws: number[];
  players: Record<string, number[]>;
}

/** rating.json: the precomputed anchors + window (same Python metric). */
export interface RatingAnchors {
  floor_metric: number;
  optimal_metric: number;
  window_gws: number[];
  captain_bonus?: boolean;
}

export interface RatingInputs {
  season: number;
  pool: RatingPoolPlayer[];
  xp: RatingXpTable;
  anchors: RatingAnchors;
}

// --------------------------------------------------------------------------------------
// Result — structurally identical to the POST /api/rate-team 200 body
// --------------------------------------------------------------------------------------

export interface RatingPlayerRow {
  player_code: number;
  web_name: string;
  position: string;
  team_short: string | null;
  price: number;
  xp_gw1: number;
  xp_horizon: number;
  q0: number | null;
  in_best_xi_gw1: boolean;
}

export interface RatingWeakestRow {
  player_code: number;
  web_name: string;
  xp_horizon: number;
}

export interface RatingScore {
  score: number;
  verdict: RatingVerdict;
  season: number;
  from_gw: number;
  horizon: number;
  team_xp_gw1: number;
  team_xp_horizon: number;
  optimal_xp_horizon: number;
  floor_xp_horizon: number;
  best_xi_gw1: number[];
  formation_gw1: string;
  suggested_captain: number;
  player_ratings: RatingPlayerRow[];
  weakest: RatingWeakestRow[];
}

export interface XiSelection {
  xi: number[];
  formation: string;
  xi_xp: number;
  captain: number;
  captain_xp: number;
}

// --------------------------------------------------------------------------------------
// Squad legality — validate_squad port (message strings verbatim)
// --------------------------------------------------------------------------------------

/** Every squad-legality rule the submitted codes violate (empty = legal). */
export function validateSquad(playerCodes: number[], pool: RatingPoolPlayer[]): string[] {
  const problems: string[] = [];
  const codes = playerCodes.map((c) => Math.trunc(c));
  if (codes.length !== SQUAD_SIZE) {
    problems.push(`need exactly ${SQUAD_SIZE} players, got ${codes.length}`);
  }
  const seen = new Map<number, number>();
  for (const code of codes) seen.set(code, (seen.get(code) ?? 0) + 1);
  const dupes = [...seen.entries()].filter(([, n]) => n > 1).map(([c]) => c);
  for (const code of dupes.sort((a, b) => a - b)) {
    problems.push(`duplicate player_code ${code}`);
  }

  const info = new Map<number, { position: string; team_code: number; price: number }>();
  for (const p of pool) {
    info.set(p.player_code, {
      position: p.position,
      team_code: p.team_code,
      price: p.price,
    });
  }
  // NB: `known` keeps duplicates, exactly like the Python list comprehension.
  const known = codes.filter((c) => info.has(c));
  const unknown = [...new Set(codes.filter((c) => !info.has(c)))];
  for (const code of unknown.sort((a, b) => a - b)) {
    problems.push(`unknown player_code ${code}`);
  }

  const counts = new Map<string, number>();
  for (const c of known) {
    const pos = info.get(c)!.position;
    counts.set(pos, (counts.get(pos) ?? 0) + 1);
  }
  for (const pos of RATING_POSITIONS) {
    if ((counts.get(pos) ?? 0) !== POSITION_QUOTA[pos]) {
      problems.push(
        `position quota: need ${POSITION_QUOTA[pos]} ${pos}, got ${counts.get(pos) ?? 0}`,
      );
    }
  }
  const clubs = new Map<number, number>();
  for (const c of known) {
    const club = info.get(c)!.team_code;
    clubs.set(club, (clubs.get(club) ?? 0) + 1);
  }
  for (const [club, n] of [...clubs.entries()].sort((a, b) => a[0] - b[0])) {
    if (n > CLUB_LIMIT) {
      problems.push(
        `club limit: max ${CLUB_LIMIT} players per club, got ${n} from team_code ${club}`,
      );
    }
  }
  let totalPrice = 0;
  for (const c of known) totalPrice += info.get(c)!.price;
  if (totalPrice > INITIAL_BUDGET) {
    problems.push(`total price ${totalPrice} exceeds budget ${INITIAL_BUDGET}`);
  }
  return problems;
}

// --------------------------------------------------------------------------------------
// The metric: greedy best XI + captain bonus, summed over the window
// --------------------------------------------------------------------------------------

/**
 * Greedy best legal XI from `[player_code, position, xp]` entries.
 *
 * Formation minimums by xP first, then filled to 11 by xP respecting the maxes.
 * Every tie breaks by player_code; the captain is the XI's highest-xP player
 * (ties -> lowest code). Throws when the entries cannot field a legal XI.
 */
export function bestXi(entries: Iterable<readonly [number, string, number]>): XiSelection {
  const byPos = new Map<string, [number, number][]>(RATING_POSITIONS.map((p) => [p, []]));
  for (const [code, pos, xp] of entries) {
    const group = byPos.get(pos);
    if (group === undefined) throw new Error(`unknown position '${pos}' for player ${code}`);
    group.push([Math.trunc(code), xp]);
  }
  for (const group of byPos.values()) {
    group.sort((a, b) => b[1] - a[1] || a[0] - b[0]); // (-xp, code)
  }

  const chosen: [number, RatingPosition, number][] = [];
  const counts: Record<string, number> = {};
  for (const pos of RATING_POSITIONS) {
    const need = FORMATION_MIN[pos];
    const group = byPos.get(pos)!;
    if (group.length < need) {
      throw new Error(`cannot field a legal XI: need ${need} ${pos}, have ${group.length}`);
    }
    for (const [code, xp] of group.slice(0, need)) chosen.push([code, pos, xp]);
    counts[pos] = need;
  }
  const rest: [number, RatingPosition, number][] = [];
  for (const pos of RATING_POSITIONS) {
    for (const [code, xp] of byPos.get(pos)!.slice(FORMATION_MIN[pos])) {
      rest.push([code, pos, xp]);
    }
  }
  rest.sort((a, b) => b[2] - a[2] || a[0] - b[0]); // (-xp, code)
  for (const [code, pos, xp] of rest) {
    if (chosen.length === XI_SIZE) break;
    if (counts[pos] >= FORMATION_MAX[pos]) continue;
    chosen.push([code, pos, xp]);
    counts[pos] += 1;
  }
  if (chosen.length < XI_SIZE) {
    throw new Error(`cannot field a legal XI: only ${chosen.length} eligible players`);
  }

  let captain = chosen[0][0];
  let captainXp = chosen[0][2];
  for (const [code, , xp] of chosen) {
    // Python max(key=(xp, -code)): higher xp wins; ties -> lower code.
    if (xp > captainXp || (xp === captainXp && code < captain)) {
      captain = code;
      captainXp = xp;
    }
  }
  const order = new Map<string, number>(RATING_POSITIONS.map((p, i) => [p, i]));
  const xi = [...chosen]
    .sort((a, b) => order.get(a[1])! - order.get(b[1])! || b[2] - a[2] || a[0] - b[0])
    .map(([code]) => code);
  let xiXp = 0;
  for (const [, , xp] of chosen) xiXp += xp;
  return {
    xi,
    formation: `${counts.DEF}-${counts.MID}-${counts.FWD}`,
    xi_xp: xiXp,
    captain,
    captain_xp: captainXp,
  };
}

/** `sum over window GWs of (best-XI xP + captain bonus)` for one 15-man squad. */
export function squadMetric(
  codes: number[],
  posOf: Map<number, string>,
  xpOf: (gw: number, code: number) => number,
  windowGws: number[],
): number {
  let total = 0;
  for (const gw of windowGws) {
    const sel = bestXi(codes.map((c) => [c, posOf.get(c)!, xpOf(gw, c)] as const));
    total += sel.xi_xp + sel.captain_xp;
  }
  return total;
}

// --------------------------------------------------------------------------------------
// rateTeam
// --------------------------------------------------------------------------------------

function clamp01(x: number): number {
  return Math.max(0, Math.min(1, x));
}

export function verdictOf(score: number): RatingVerdict {
  for (const [threshold, verdict] of VERDICT_BANDS) {
    if (score >= threshold) return verdict;
  }
  return "FODDER";
}

/**
 * Score a 15-man squad 0-100 between the floor and dream-team anchors — the
 * Python rate_team, fed from the static bundle. Pure and deterministic; players
 * with no xp.json row contribute 0 xP but remain legal picks. Throws
 * RatingValidationError with EVERY violated legality rule for an illegal squad.
 */
export function rateTeam(playerCodes: number[], inputs: RatingInputs): RatingScore {
  const { pool, xp, anchors } = inputs;
  const problems = validateSquad(playerCodes, pool);
  if (problems.length > 0) throw new RatingValidationError(problems);

  const codes = playerCodes.map((c) => Math.trunc(c));
  const windowGws = [...anchors.window_gws].sort((a, b) => a - b);
  const fromGw = windowGws[0];
  const horizon = windowGws.length > 0 ? windowGws[windowGws.length - 1] - fromGw + 1 : 0;

  const gwIndex = new Map<number, number>(xp.gws.map((g, i) => [g, i]));
  const xpOf = (gw: number, code: number): number => {
    const i = gwIndex.get(gw);
    if (i === undefined) return 0;
    return xp.players[String(code)]?.[i] ?? 0;
  };

  const posOf = new Map<number, string>();
  const byCode = new Map<number, RatingPoolPlayer>();
  for (const p of pool) {
    posOf.set(p.player_code, p.position);
    byCode.set(p.player_code, p);
  }

  const teamMetric = squadMetric(codes, posOf, xpOf, windowGws);
  const optimalMetric = anchors.optimal_metric;
  const floorMetric = anchors.floor_metric;

  const denom = optimalMetric - floorMetric;
  let score: number;
  if (denom > 0) {
    score = clamp01((teamMetric - floorMetric) / denom) * 100;
  } else {
    // Degenerate anchors (e.g. empty window): pass/fail split, as in Python.
    score = teamMetric >= optimalMetric ? 100 : 0;
  }

  const selGw1 = bestXi(codes.map((c) => [c, posOf.get(c)!, xpOf(fromGw, c)] as const));
  const xiGw1 = new Set(selGw1.xi);

  const xpHorizonOf = new Map<number, number>();
  for (const c of codes) {
    let sum = 0;
    for (const gw of windowGws) sum += xpOf(gw, c);
    xpHorizonOf.set(c, sum);
  }

  const ratings: RatingPlayerRow[] = codes.map((c) => {
    const p = byCode.get(c)!;
    return {
      player_code: c,
      web_name: p.web_name,
      position: p.position,
      team_short: p.team_short ?? null,
      price: p.price,
      xp_gw1: xpOf(fromGw, c),
      xp_horizon: xpHorizonOf.get(c)!,
      q0: p.q0_gw1 ?? null,
      in_best_xi_gw1: xiGw1.has(c),
    };
  });
  const weakest: RatingWeakestRow[] = [...codes]
    .sort((a, b) => xpHorizonOf.get(a)! - xpHorizonOf.get(b)! || a - b)
    .slice(0, 3)
    .map((c) => ({
      player_code: c,
      web_name: byCode.get(c)!.web_name,
      xp_horizon: xpHorizonOf.get(c)!,
    }));

  return {
    score,
    verdict: verdictOf(score),
    season: inputs.season,
    from_gw: fromGw,
    horizon,
    team_xp_gw1: selGw1.xi_xp + selGw1.captain_xp,
    team_xp_horizon: teamMetric,
    optimal_xp_horizon: optimalMetric,
    floor_xp_horizon: floorMetric,
    best_xi_gw1: selGw1.xi,
    formation_gw1: selGw1.formation,
    suggested_captain: selGw1.captain,
    player_ratings: ratings,
    weakest,
  };
}
